x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SuccessfulAuthenticationResponseMessageBuilder.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.idp.saml.authn;
+
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.xpack.idp.saml.idp.SamlIdentityProvider;
+import org.elasticsearch.xpack.idp.saml.sp.SamlServiceProvider;
+import org.elasticsearch.xpack.idp.saml.support.SamlFactory;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.opensaml.core.xml.schema.XSString;
+import org.opensaml.saml.saml2.core.Assertion;
+import org.opensaml.saml.saml2.core.Attribute;
+import org.opensaml.saml.saml2.core.AttributeStatement;
+import org.opensaml.saml.saml2.core.AttributeValue;
+import org.opensaml.saml.saml2.core.Audience;
+import org.opensaml.saml.saml2.core.AudienceRestriction;
+import org.opensaml.saml.saml2.core.AuthnRequest;
+import org.opensaml.saml.saml2.core.Conditions;
+import org.opensaml.saml.saml2.core.Issuer;
+import org.opensaml.saml.saml2.core.NameID;
+import org.opensaml.saml.saml2.core.NameIDType;
+import org.opensaml.saml.saml2.core.Response;
+import org.opensaml.saml.saml2.core.Status;
+import org.opensaml.saml.saml2.core.StatusCode;
+import org.opensaml.saml.saml2.core.Subject;
+
+import java.time.Clock;
+import java.util.Collection;
+
+/**
+ * Builds SAML 2.0 {@link Response} objects for successful authentication results.
+ */
+public class SuccessfulAuthenticationResponseMessageBuilder {
+
+    private final SamlFactory samlFactory;
+    private final Clock clock;
+    private final SamlIdentityProvider idp;
+
+    public SuccessfulAuthenticationResponseMessageBuilder(SamlFactory samlFactory, Clock clock, SamlIdentityProvider idp) {
+        this.samlFactory = samlFactory;
+        this.clock = clock;
+        this.idp = idp;
+    }
+
+    public Response build(UserServiceAuthentication user, @Nullable AuthnRequest request) {
+        final DateTime now = now();
+        final SamlServiceProvider serviceProvider = user.getServiceProvider();
+
+        final Response response = samlFactory.object(Response.class, Response.DEFAULT_ELEMENT_NAME);
+        response.setID(samlFactory.secureIdentifier());
+        if (request != null) {
+            response.setInResponseTo(request.getID());
+        }
+        response.setIssuer(buildIssuer());
+        response.setIssueInstant(now);
+        response.setStatus(buildStatus());
+        response.setDestination(serviceProvider.getAssertionConsumerService().toString());
+
+        final Assertion assertion = samlFactory.object(Assertion.class, Assertion.DEFAULT_ELEMENT_NAME);
+        assertion.setID(samlFactory.secureIdentifier());
+        assertion.setIssuer(buildIssuer());
+        assertion.setIssueInstant(now);
+        assertion.setConditions(buildConditions(now, serviceProvider));
```

Comment:
Why didn't the schema pick me up on that... :(

Reply:
This is a requirement for the Web browser SSO profile so it's not covered by the Assertion schema

Reply:
Ah, because Saml2Int expects it (as do we) but the schema doesn't.

Reply:
Not that it matters too much but it's not saml2int but rather the actual OASIS standard [4.1.4.2](https://docs.oasis-open.org/security/saml/v2.0/saml-profiles-2.0-os.pdf). Only mentioning as it can be argued that saml2int is not mandatory to follow ( but rather a choice ), but the actual spec is.